tests/test_compare_dictconfig_vs_dict.py

```diff
@@ -0,0 +1,647 @@
+    TestEnumTypeDunderMethods: for DictConfig where key_type is Enum
+"""
+from copy import deepcopy
+from enum import Enum
+from typing import Any, Dict, Optional
+
+from pytest import fixture, mark, param, raises
+
+from omegaconf import DictConfig, OmegaConf
+from omegaconf.errors import ConfigKeyError, ConfigTypeError, KeyValidationError
+from tests import Enum1
+
+
+@fixture(
+    params=[
+        "str",
+        1,
+        3.1415,
+        True,
+        Enum1.FOO,
+    ]
+)
+def key(request: Any) -> Any:
+    """A key to test indexing into DictConfig."""
+    return request.param
+
+
+@fixture
+def python_dict(data: Dict[Any, Any]) -> Dict[Any, Any]:
+    """Just a standard python dictionary, to be used in comparison with DictConfig."""
+    return deepcopy(data)
+
+
+@fixture(params=[True, False, None])
+def struct_mode(request: Any) -> Optional[bool]:
+    struct_mode: Optional[bool] = request.param
+    return struct_mode
+
+
+@mark.parametrize(
+    "data",
+    [
+        param({"a": 10}, id="str"),
+        param({1: "a"}, id="int"),
+        param({123.45: "a"}, id="float"),
+        param({True: "a"}, id="bool"),
+        param({Enum1.FOO: "foo"}, id="Enum1"),
+    ],
+)
+class TestUntypedDictConfig:
+    """Compare DictConfig with python dict in the case where key_type is not set."""
+
+    @fixture
+    def cfg(self, python_dict: Any, struct_mode: Optional[bool]) -> DictConfig:
+        """Create a DictConfig instance from the given data"""
+        cfg: DictConfig = DictConfig(content=python_dict)
+        OmegaConf.set_struct(cfg, struct_mode)
+        return cfg
+
+    def test__setitem__(
+        self, python_dict: Any, cfg: DictConfig, key: Any, struct_mode: Optional[bool]
+    ) -> None:
+        """Ensure that __setitem__ has same effect on python dict and on DictConfig."""
+        if struct_mode and key not in cfg:
+            with raises(ConfigKeyError):
+                cfg[key] = "sentinel"
+        else:
+            python_dict[key] = "sentinel"
+            cfg[key] = "sentinel"
+            assert python_dict == cfg
+
+    def test__getitem__(self, python_dict: Any, cfg: DictConfig, key: Any) -> None:
+        """Ensure that __getitem__ has same result with python dict as with DictConfig."""
+        try:
+            result = python_dict[key]
+        except KeyError:
+            with raises(ConfigKeyError):
+                cfg[key]
+        else:
+            assert result == cfg[key]
+
+    @mark.parametrize("struct_mode", [False, None])
+    def test__delitem__(self, python_dict: Any, cfg: DictConfig, key: Any) -> None:
+        """Ensure that __delitem__ has same result with python dict as with DictConfig."""
+        try:
+            del python_dict[key]
+            assert key not in python_dict
+        except KeyError:
+            with raises(ConfigKeyError):
+                del cfg[key]
+        else:
+            del cfg[key]
+            assert key not in cfg
```

Comment:
Tests for non-struct behaviors are marked with `@mark.parametrize("struct_mode", [False, None])`, which will override the default values for `struct_mode in [True, False, None]`.

Reply:
I would differentiate the case for a key that is found and a key that is not found into two tests.
This probably means that it's easier to not use the key fixture at all here.
```python
@mark.parameterize(("cfg", "key"), [
  ({"key": 1}, "key"),
  ({True: 1}, True),
  ...
])
def test__delitem__found(cfg, key):
  del cfg[key]
  assert key not in cfg

@mark.parameterize(("cfg", "key"), [
  ({"key": 1}, "key"),
  ({True: 1}, True),
  ...
])
def test__delitem__not_found(cfg):
  with raises(KeyError):
    del cfg[key]
```

I don't think we need to worry about testing that the exception type is ConfigKeyError here, we just care about similarity to plain dict.

This can be a part of the followup reffactor of this class.